src/editor/Core/Filesystem.re

```diff
@@ -0,0 +1,301 @@
+   value is passed to the success handler
+
+   The advantage is that it allows for a concise way to describe a whole
+   range of error and OK scenarios
+ */
+let (>>=) = onSuccess;
+let (/\/=) = onError;
+/* let ( *> ) = always; */
+
+/**
+   Permissions ==================================================
+ */
+let userReadWriteExecute = 0o777;
+
+/**
+   Safe Unix functions ==========================================
+ */
+
+let isFile = st =>
+  switch (st.Unix.st_kind) {
+  | Unix.S_REG => return()
+  | _ => error("not a file")
+  };
+
+let isDir = st =>
+  switch (st.Unix.st_kind) {
+  | Unix.S_DIR => return()
+  | _ => error("not a directory")
+  };
+
+let hasOwner = (uid, st) =>
+  if (st.Unix.st_uid == uid) {
+    return();
+  } else {
+    error("expected uid = %d, found %d", uid, st.Unix.st_uid);
+  };
+
+let hasGroup = (gid, st) =>
+  if (st.Unix.st_gid == gid) {
+    return();
+  } else {
+    error("expected gid = %d, found %d", gid, st.Unix.st_gid);
+  };
+
+let hasPerm = (perm, st) =>
+  if (st.Unix.st_perm == perm) {
+    return();
+  } else {
+    error("expected permissions 0o%o, found 0o%o", perm, st.Unix.st_perm);
+  };
+
+let getGroupId = () =>
+  Unix.(
+    try (getgid() |> return) {
+    | Not_found => error("No Group ID found")
+    }
+  );
+
+let getUserId = () =>
+  Unix.(
+    try (getuid() |> return) {
+    | Not_found => error("Cannot find user")
+    }
+  );
+
+let stat = path =>
+  Unix.(
+    try (Some(stat(path)) |> return) {
+    | Unix_error(ENOENT, _, _) => return(None)
+    }
+  );
+
+let chmod = (path, ~perm=userReadWriteExecute, ()) =>
+  Unix.(
+    try (chmod(path, perm) |> return) {
+    | Unix_error(_, _, _) => error("can't set permissions for '%s'", path)
+    }
+  );
+
+let chown = (path, uid, gid) =>
+  Unix.(
+    try (chown(path, uid, gid) |> return) {
+    | Unix_error(err, _, _) =>
+      error(
+        "can't set uid/gid for '%s' because '%s",
+        path,
+        error_message(err),
+      )
+    }
+  );
+
+/**
+  Permissions default to read/write permissions
+
+  The permissions of the directory should ideally be
+  drwxr-xr-x -> this means the user can read, write and execute
+  but group and others can only read and execute. Without
+  the correct permission creating the subfolders or files
+  will not work correctly
+*/
+let mkdir = (path, ~perm=userReadWriteExecute, ()) =>
+  Unix.(
+    try (mkdir(path, perm) |> return) {
+    | Unix_error(err, _, _) =>
+      error(
+        "can't create directory '%s' because '%s",
+        path,
+        error_message(err),
+      )
+    }
+  );
+
+let rmdir = path =>
+  Unix.(
+    try (rmdir(path) |> return) {
+    | Unix_error(err, _, _) =>
+      error(
+        "can't remove directory '%s' because: '%s'",
+        path,
+        error_message(err),
+      )
+    }
+  );
+
+let getOniDirectory = home =>
+  Utility.join([home, ".config", "oni2"]) |> return;
+
+let getHomeDirectory = () =>
+  Unix.(
+    try (getenv("HOME") |> return) {
```

Comment:
Ah, this might not be sufficient on all platforms - here's an example of a check that works for Windows too from `esy`: https://github.com/esy/esy/blob/b0ea114219c77855e0fb36e3cdcfae72adad31b4/esy-lib/Path.re#L36

Reply:
Do we want to store things in `home` for Windows?

The call a few lines down in that file is probably what we want, since `%APPDATA%` is the common store for config and whatnot on Windows. `%APPDATA%/oni` is where Oni1 stored stuff etc.

Reply:
Good point @CrossR . %APPDATA%/oni2 seems like the right spot on Windows for config.

Reply:
:man_facepalming: I just changed it to home, :laughing: I'll have another look at using the data path for windows

Reply:
Fixed :+1: